vllm/model_executor/layers/rotary_embedding.py

```diff
@@ -75,6 +71,27 @@ def _apply_rotary_emb(
         return torch.stack((o1, o2), dim=-1).flatten(-2)
 
 
+def _apply_rotary_emb(x: torch.Tensor,
+                      cos: torch.Tensor,
+                      sin: torch.Tensor,
+                      is_neox_style: bool,
+                      use_flash_attn=False) -> torch.Tensor:
+    """
+    Args:
+        x: [num_tokens, num_heads, head_size]
+        cos: [num_tokens, head_size // 2]
+        sin: [num_tokens, head_size // 2]
+        is_neox_style: Whether to use the Neox-style or GPT-J-style rotary
+            positional embeddings.
+        use_flash_attn: Whether to enable Flash Attention optimizations.
+    """
+    if use_flash_attn:
+        return apply_rotary_emb(x.unsqueeze(0), cos, sin,
+                                not is_neox_style).squeeze(0)
+    else:
+        return _apply_rotary_emb_torch(x, cos, sin, is_neox_style)
```

Comment:
```suggestion
    if current_platform.is_cuda_alike():
        from vllm_flash_attn.layers.rotary import apply_rotary_emb
        return apply_rotary_emb(x.unsqueeze(0), cos, sin,
                                not is_neox_style).squeeze(0)
    else:
        return _apply_rotary_emb_torch(x, cos, sin, is_neox_style)
```

Since triton is only available for Nvidia and ROCm GPUs, we can simplify the implementation here.

Reply:
hi，isotr0py，if I replace
```
        from flash_attn.layers.rotary import apply_rotary_emb
        return apply_rotary_emb(x.unsqueeze(0), cos, sin,
                                not is_neox_style).squeeze(0)
```
to
```
        from vllm_flash_attn.layers.rotary import apply_rotary_emb
        return apply_rotary_emb(x.unsqueeze(0), cos, sin,
                                not is_neox_style).squeeze(0)
```
Running the CI test will result in an error：
https://buildkite.com/vllm/ci/builds/18001#01964d1c-6acb-48b6-8b8e-6b296cdefd17

[2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/qwen2.py", line 243, in forward
--
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     hidden_states = self.self_attn(
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]                     ^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     return self._call_impl(*args, **kwargs)
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     return forward_call(*args, **kwargs)
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/qwen2.py", line 176, in forward
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     q, k = self.rotary_emb(positions, q, k)
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     return self._call_impl(*args, **kwargs)
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     return forward_call(*args, **kwargs)
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/rotary_embedding.py", line 992, in forward
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     query_rot = _apply_rotary_emb(query_rot, cos, sin, self.is_neox_style)
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/rotary_embedding.py", line 81, in _apply_rotary_emb
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386]     from vllm_flash_attn.layers.rotary import apply_rotary_emb
  | [2025-04-19T08:53:45Z] ERROR 04-19 01:53:45 [core.py:386] ModuleNotFoundError: No module named 'vllm_flash_attn'

Reply:
I use flash.attn.layers.gross import
Mainly because qwen2-vl.py is used in this way, as follows
```
def apply_rotary_pos_emb_vision(t: torch.Tensor,
                                freqs: torch.Tensor,
                                use_flash_attn=False) -> torch.Tensor:
    t_ = t.float()
    cos = freqs.cos()
    sin = freqs.sin()
    apply_rotary_emb = apply_rotary_emb_torch
    if use_flash_attn:
        from flash_attn.layers.rotary import apply_rotary_emb
    output = apply_rotary_emb(t_, cos, sin).type_as(t)
    return output
```

Reply:
Oh, the import code should be `vllm.vllm_flash_attn.layers.rotary import apply_rotary_emb`.

Qwen2-vl use original FA due to `vllm_flash_attn` compatibility issue for ViT, however, we shouldn't use it here since original FA is not a requirement.

Reply:
Using vllm-vllm_flash_mattn.layers.total still encounters errors：

  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]     return forward_call(*args, **kwargs)
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/rotary_embedding.py", line 992, in forward
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]     query_rot = _apply_rotary_emb(query_rot, cos, sin, self.is_neox_style)
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/rotary_embedding.py", line 81, in _apply_rotary_emb
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386]     from vllm.vllm_flash_attn.layers.rotary import apply_rotary_emb
  | [2025-04-19T15:42:15Z] ERROR 04-19 08:42:15 [core.py:386] ModuleNotFoundError: No module named 'vllm.vllm_flash_attn.layers'

I tried it_ The rotary_imbedding in custom_ops-py can be used

Reply:
Hmmm, seems `vllm_flash_attn` only ports the FA interface (https://github.com/vllm-project/flash-attention/tree/main/vllm_flash_attn), we might need to fork the triton rotary kernels manually under the `layer` folder...

Reply:
Are you saying that I need to make modifications https://github.com/vllm-project/flash-attention project ？
Can I modify it like this:
1. Create a 'layers' sub directory in the' vllm_flash-attn 'directory
2. Then copy 'flash_attn/layers/total. py' to the 'vllm_flash-attn/layers/' directory,
3. Import or export 'apply_rotary-emb' in 'vllm_flash-attn/layers/__init__. py'.
4. Ensure that the top-level '__init__. py' of 'vllm_flash-attn' correctly exports the 'layers' module.

Reply:
> Are you saying that I need to make modifications https://github.com/vllm-project/flash-attention project ？

Yes. We should not directly copy the source code from FA to `vllm_flash_attn`, because the copying should happen at the compilation during installation. So I recommend to make modifications in https://github.com/vllm-project/flash-attention to make rotary code copied during compilation.

Reply:
The file I want to copy,/flash attention/flash attn/layers/total. py, contains the following code
`from flash_attn.ops.triton.rotary import apply_rotary
`
If I copy  /flash-attention/flash-attn/layers/rotary.py to the /flash-attention/vllm_flash-attn/ directory during compilation, the error will still occur during execution

Reply:
> > Are you saying that I need to make modifications https://github.com/vllm-project/flash-attention project ？
> 
> Yes. We should not directly copy the source code from FA to `vllm_flash_attn`, because the copying should happen at the compilation during installation. So I recommend to make modifications in https://github.com/vllm-project/flash-attention to make rotary code copied during compilation.

Can you help me merge this flash attn PR？
https://github.com/vllm-project/flash-attention/pull/64